# Ejercicio 3: Modelo vectorial y TF-IDF
### Nombre: Anthony Goyes
### Paso 1: Calcular la matriz TF-IDF para el corpus data/01_corpus_turismo_500.txt

En este paso se carga el archivo `01_corpus_turismo_500.txt`, donde cada línea representa un documento.

Se utiliza TF-IDF para convertir los documentos en vectores numéricos.

In [20]:
from sklearn.feature_extraction.text import TfidfVectorizer

ruta_turismo = "../data/01_corpus_turismo_500.txt"

with open(ruta_turismo, "r", encoding="utf-8") as f:
    documentos_turismo = f.readlines()

vectorizer_turismo = TfidfVectorizer()
matriz_tfidf_turismo = vectorizer_turismo.fit_transform(documentos_turismo)

print("Documentos turismo:", len(documentos_turismo))
print("Matriz TF-IDF turismo:", matriz_tfidf_turismo.shape)

Documentos turismo: 500
Matriz TF-IDF turismo: (500, 116)


La matriz TF-IDF obtenida tiene dimensiones (500, 116), lo que significa:

- 500 documentos (filas)
- 116 términos distintos (columnas)

Cada valor representa la importancia de un término dentro de un documento.

### Paso 2: Construir el corpus Gutenberg 1000

Se verifica que el corpus contiene 1000 libros en formato texto.

In [21]:
import os

ruta_gutenberg = "../data/gutenberg/gutenberg_1000"

archivos = [f for f in os.listdir(ruta_gutenberg) if f.endswith(".txt")]

print("Cantidad de libros:", len(archivos))

Cantidad de libros: 1000


### Paso 3: Calcular la matriz TF-IDF para el corpus Gutenberg 1000

Se cargan los 1000 libros y se calcula su representación vectorial usando TF-IDF.

Se limita a 10000 términos y se eliminan palabras comunes en inglés.

In [22]:
corpus_gutenberg = []
nombres_archivos = []

for archivo in sorted(archivos):
    ruta = os.path.join(ruta_gutenberg, archivo)
    
    with open(ruta, "r", encoding="utf-8", errors="ignore") as f:
        texto = f.read()
        corpus_gutenberg.append(texto)
        nombres_archivos.append(archivo)

vectorizer_gutenberg = TfidfVectorizer(
    max_features=10000,
    stop_words="english"
)

matriz_tfidf_gutenberg = vectorizer_gutenberg.fit_transform(corpus_gutenberg)

print("Documentos Gutenberg:", len(corpus_gutenberg))
print("Matriz TF-IDF Gutenberg:", matriz_tfidf_gutenberg.shape)

Documentos Gutenberg: 1000
Matriz TF-IDF Gutenberg: (1000, 10000)


La matriz TF-IDF tiene dimensiones (1000, 10000), lo que indica:

- 1000 documentos (libros)
- 10000 términos más relevantes del corpus

Se utilizaron stopwords en inglés para mejorar la calidad de los resultados.

### Paso 4: Programar una función buscar() para el corpus Gutenberg 1000

Se implementa una función que recibe una consulta y devuelve los documentos más relevantes utilizando similitud coseno.

In [23]:
from sklearn.metrics.pairwise import cosine_similarity

def buscar(consulta, top_k=5):
    consulta_vector = vectorizer_gutenberg.transform([consulta])
    similitudes = cosine_similarity(consulta_vector, matriz_tfidf_gutenberg).flatten()
    
    indices = similitudes.argsort()[::-1][:top_k]
    
    for i in indices:
        print("Archivo:", nombres_archivos[i])
        print("Similitud:", similitudes[i])
        print("-" * 40)

### Ejemplo de búsqueda

Consulta: "love marriage family"

Se muestran los documentos más similares según la similitud coseno.

In [24]:
buscar("love marriage family", top_k=5)

Archivo: book_0376.txt
Similitud: 0.2042077630820382
----------------------------------------
Archivo: book_0572.txt
Similitud: 0.17618779701862722
----------------------------------------
Archivo: book_0418.txt
Similitud: 0.16988609759188925
----------------------------------------
Archivo: book_0762.txt
Similitud: 0.16249596056837592
----------------------------------------
Archivo: book_0445.txt
Similitud: 0.15362399937071325
----------------------------------------
